### PR TITLE
Dead peers

### LINKED ISSUE #1070: Add extra check to dead peer removal, that pings the proposed dead peer to verify a timeout. In the case that the node actually isn't down, create an Observation for "IncorrectDeadPeerNotice".

### PR DESCRIPTION
**Pull request by marcinwadon**
Resolves #1070

### CODE REVIEW

File: src/main/scala/org/constellation/domain/p2p/PeerHealthCheck.scala [hunk 1]
```diff
@@ -1,61 +1,101 @@
...
+    peers.traverse { pd =>
+      for {
+        _ <- logger.debug(
+          s"Waiting for network confirmation before marking a dead peer as offline: ${pd.peerMetadata.id.short} (${pd.peerMetadata.host})"
+        )
+        statuses <- cluster.broadcast[PeerHealthCheckStatus](
+          apiClient.cluster.checkPeerResponsiveness(pd.peerMetadata.id).run,
+          Set(pd.peerMetadata.id)
+        )
+        allUnresponsive = statuses.mapValues(_.fold(_ => PeerUnresponsive, identity)).toSeq.map(_._2).toSet.size == 1
```
Comment (TheMMaciek) [inline src/main/scala/org/constellation/domain/p2p/PeerHealthCheck.scala:85]: Won't this piece of code indicate a peer as unresponsive incorrectly if all responses are actually PeerAvailable?
The check at the and could be == Set(PeerUnresponsive)

Comment (marcinwadon) [reply src/main/scala/org/constellation/domain/p2p/PeerHealthCheck.scala:85]: :+1:

File: src/main/scala/org/constellation/domain/p2p/PeerHealthCheck.scala [hunk 2]
```diff
@@ -1,61 +1,109 @@
...
+      for {
+        _ <- logger.debug(
+          s"Waiting for network confirmation before marking a dead peer as offline: ${pd.peerMetadata.id.short} (${pd.peerMetadata.host})"
+        )
+        statuses <- cluster.broadcast[PeerHealthCheckStatus](
+          apiClient.cluster.checkPeerResponsiveness(pd.peerMetadata.id).run,
+          Set(pd.peerMetadata.id)
+        )
+        atLeastOneAvailable = statuses
+          .mapValues(_.fold(_ => PeerUnresponsive, identity))
```
Comment (kpudlik) [inline src/main/scala/org/constellation/domain/p2p/PeerHealthCheck.scala:90]: If you `fold` with `identity` for right side of `Either` then very likely you just need `left` projection as you want to touch left side only :

```scala
.mapValues(_.left.map(_ => PeerUnresponsive))
```

File: src/main/scala/org/constellation/domain/p2p/PeerHealthCheck.scala [hunk 3]
```diff
@@ -1,61 +1,109 @@
...
+          .toSeq
+          .map(_._2)
```
Comment (kpudlik) [inline src/main/scala/org/constellation/domain/p2p/PeerHealthCheck.scala:92]: Just make `.values`

File: src/main/scala/org/constellation/domain/p2p/PeerHealthCheck.scala (hunk 2 above)
Comment (kpudlik) [reply src/main/scala/org/constellation/domain/p2p/PeerHealthCheck.scala:90]: Actually that step looks redundant ad you later do `contains(PeerAvailable)` on right side. You never check left side.

File: src/main/scala/org/constellation/domain/p2p/PeerHealthCheck.scala (hunk 3 above)
Comment (kpudlik) [reply src/main/scala/org/constellation/domain/p2p/PeerHealthCheck.scala:92]: Or even do:

```
statuses
   .mapValues(_.fold(_ => PeerUnresponsive, identity))
   .exists { case(_ status) => status.contains(PeerAvailable)
```

File: src/main/scala/org/constellation/domain/p2p/PeerHealthCheck.scala (hunk 2 above)
Comment (kpudlik) [inline src/main/scala/org/constellation/domain/p2p/PeerHealthCheck.scala:90]: Eventually:

```scala
.mapValues(_.getOrElse(PeerUnresponsive))
```

File: src/main/scala/org/constellation/domain/p2p/PeerHealthCheck.scala [hunk 4]
```diff
@@ -1,61 +1,109 @@
...
       unresponsivePeers = statuses.count(_._2 == PeerUnresponsive)
-      _ <- if (unresponsivePeers > 0) logger.info(s"Found dead peers: ${unresponsivePeers}") else Sync[F].unit
-      _ <- markOffline(statuses)
+      _ <- if (unresponsivePeers > 0) logger.info(s"Found dead peers: ${unresponsivePeers}. Ensuring offline")
+      else Sync[F].unit
+      unresponsive <- statuses.filter { case (_, status) => status == PeerUnresponsive }.filterA {
+        case (pd, _) => ensureOffline(pd.peerMetadata.toPeerClientMetadata)
+      }
+      offlinePeers <- confirmUnresponsivePeers(unresponsive.map(_._1))
+      _ <- markOffline(offlinePeers)
```
[the remark this answers is not in the document]
Comment (marcinwadon) [reply src/main/scala/org/constellation/domain/p2p/PeerHealthCheck.scala]: :+1:

**Pull request merged**